src/main/java/tech/sourced/siva/IndexReader.java

```diff
@@ -89,55 +108,125 @@ private Index readIndex(final BaseIndex index) throws SivaException {
 
-        FileTime modificationTime = FileTime.from(this.sivaFile.readLong(), TimeUnit.NANOSECONDS);
+        int rawFileMode = this.sivaFile.readInt();
+        Set<PosixFilePermission> fileMode = FileModeUtils.posixFilePermissions(rawFileMode);
+
+        long rawModTime = this.sivaFile.readLong();
+        FileTime modificationTime = FileTime.from(rawModTime, TimeUnit.NANOSECONDS);
 
         long fileOffset = this.sivaFile.readLong();
+        checkUnsignedLongs(fileOffset, "At Index Entry " + name + ", file offset: ");
+
         long fileSize = this.sivaFile.readLong();
-        int crc32 = this.sivaFile.readInt();
+        checkUnsignedLongs(fileSize, "At Index Entry " + name + ", file size: ");
+
+        int rawCrc32 = this.sivaFile.readInt();
+        long crc32 = castUnsignedIntToLong(rawCrc32);
 
-        Flag flag = Flag.fromInteger(this.sivaFile.readInt());
+        int rawFlag = this.sivaFile.readInt();
+        Flag flag = Flag.fromInteger(rawFlag);
+
+        long beginOfEntry = (endOfBlock - indexFooter.getBlockSize()) + fileOffset;
 
         return new IndexEntry(
-                new String(name),
+                name,
+                fileMode,
                 modificationTime,
-                FileModeUtils.posixFilePermissions(fileMode),
                 flag,
                 fileOffset,
                 fileSize,
                 crc32,
-                (block - indexFooter.getBlockSize()) + fileOffset
+                beginOfEntry
         );
     }
 
-    private IndexFooter readIndexFooter() throws IOException {
-        return new IndexFooter(
-                this.sivaFile.readInt(),
-                this.sivaFile.readLong(),
-                this.sivaFile.readLong(),
-                this.sivaFile.readInt()
-        );
+    private long castUnsignedIntToLong(final int n) {
+        return n & READ_UINT32_MASK;
     }
 
-    private void readIndexVersion() throws IOException, SivaException {
-        int version = this.sivaFile.readByte();
-        if (version != INDEX_VERSION) {
-            throw new SivaException("Invalid index version at " + this.sivaFileName + " file.");
+    private void checkUnsignedLongs(final long n, final String from) throws SivaException {
+        if (n < 0) {
+            throw new SivaException(from + SivaException.UNSIGNED_LONG);
         }
     }
 
-    private void readSignature() throws IOException, SivaException {
-        byte[] sig = new byte[INDEX_SIGNATURE.length];
-        this.sivaFile.readFully(sig);
-        if (!Arrays.equals(sig, INDEX_SIGNATURE)) {
-            throw new SivaException("Invalid index signature at " + this.sivaFileName + " file.");
+    private void checkIndexCrc(final IndexFooter footer, final long startOfIndex)
+            throws IOException, SivaException {
+
+        // The checksum object could be fed with more than (2^31)-1 bytes
+        // wich is the most length for a data structure the JVM allows.
```

Comment:
s/wich/which/ s/most length for a data structure the JVM allows/biggest length allowed by the JVM for a data structure/